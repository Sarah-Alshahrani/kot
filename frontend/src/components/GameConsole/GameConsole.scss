.game_console {}

.submit_game_command {
  color: black;
}

$blue: #0084ff;
$white: #fff;
$grey: #eee;

.chat {
  // margin: auto;

  .container {
    // padding: 0 30px;
    // height: 80%;
    // width: 40%;
    height: 50vh;

    &.message-form {
      // height: 10%;
      margin-top: 1px;
      margin-bottom: 3px;
    }
  }

  h1,
  h3 {
    margin-top: 0;
    color: $blue;
  }

  input {
    color: #7e99a0;
    font-size: 1em;
    font-weight: 300;
    // line-height: 19px;
    padding: 5px;
    margin-right: 5px;
    box-sizing: border-box;
    display: inline-block;
    width: 75%;
    float: left;
  }

  button {
    background-color: $blue;
    border-color: $blue;
    border-style: solid;
    color: white;
    font-size: 1em;
    // padding: 5px;
    width: 20%;
    // float: left;
  }

  ul {
    list-style: none;
    overflow-y: auto;
    margin: 0;
    padding: 0;
    height: 90%;
    width: 100%;
  }

  ul li {
    clear: both;
    display: block;
    margin-bottom: 5px;
    // max-width: 80%;
    white-space: normal;
    float: left;
    width: 100%;
    height: auto;
    word-wrap: break-word;

    p {
      border-radius: 100px / 30px;
      padding: 5px 20px;
      margin: 0;
    }

    .author {
      color: grey;
      display: inline-block;
      text-align: center;
      padding: 5px;
      margin: 0 2%;
      font-weight: 300;
      font-size: 12px;
    }
  }

  .him {
    float: left;
    text-align: left;

    p {
      background: $white;
    }
  }

  .me {
    float: right;
    text-align: right;

    .author {
      display: none;
    }

    p {
      background: $blue;
      color: $white;
    }
  }

  .him+.me {
    p {
      border-bottom-right-radius: 1px;
    }
  }

  .me+.me {
    p {
      border-top-right-radius: 1px;
      border-bottom-right-radius: 1px;
    }
  }

  .me:last-of-type {
    p {
      border-bottom-right-radius: 10px;
    }
  }
}