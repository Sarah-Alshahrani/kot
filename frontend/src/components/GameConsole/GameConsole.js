import React from "react";
import "./GameConsole.css";

import GameLog from "./GameLog";

import GameInstance from "./../../services/gameService";

const gameLogExample = [
  // '-- Sammy turn â€“',
  // 'Rolled: E E H 1 2 3',
  // 'Decision: keep E E',
  // 'Rolled: E P 1 2 2',
  // 'Decision: keep E',
  // 'Rolled: 2 2 2 (TRIPPLE! +2 VP)',
  // ':: Gain: E E E +2 VP ::',
  // 'Buy: Psychic Probe',
  // 'End Turn'
];

class GameConsole extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      username: props.currentUser,
      gameRoom: props.currentRoom,
      log: [], // this holds the name of each list
      cmd: "enter cmd"
    };

    this.waitForSocketConnection(() => {
      const userObject = {
        user: this.state.username,
        room: this.state.gameRoom,
        data: this.state.cmd
      };

      GameInstance.initUser(userObject);
      GameInstance.addCallbacks(this.serverResponseHandler.bind(this));
    });

    this.handleChange = this.handleChange.bind(this);
    this.SubmitGameCommand = this.SubmitGameCommand.bind(this);
  }

  serverResponseHandler(messages) {
    this.appendStateCmdToLog(messages.content);
  }

  waitForSocketConnection(callback) {
    const component = this;
    setTimeout(function() {
      // Check if websocket state is OPEN
      if (GameInstance.state() === 1) {
        console.log("Connection is made");
        callback();
        return;
      } else {
        console.log("wait for connection...");
        component.waitForSocketConnection(callback);
      }
    }, 100); // wait 100 milisecond for the connection...
  }

  appendStateCmdToLog(message) {
    const action = message.actionType;
    // const room = message.room;
    // const user = message.user;
    const content = message.content;

    if (content === undefined || content == "") return;

    switch (action) {
      case "GAMECONSOLE_ERROR":
        break;
      case "GAMECONSOLE_MESSAGE":
        break;
      case "GAMECONSOLE_DECISION":
        break;

      default:
        return;
    }

    this.setState({ log: [...this.state.log, content] });

    return this.state.log;
  }

  SubmitGameCommand(e) {
    const cmd = {
      user: this.state.username,
      room: this.state.gameRoom,
      content: this.state.cmd
    };

    this.appendStateCmdToLog(cmd);

    const messageObject = {
      user: this.state.username,
      room: this.state.gameRoom,
      data: this.state.cmd
    };

    GameInstance.sendGameLogCommand(messageObject);
  }

  handleChange(e) {
    this.setState({
      cmd: e.target.value
    });
  }

  render() {
    return (
      <div className="game_console">
        <GameLog
          data={this.state.log}
          currentUser={this.state.username}
          currentRoom={this.state.gameRoom}
        />

        <form
          className="submit_game_command"
          onSubmit={this.SubmitGameCommand}
          onChange={this.handleChange}
        >
          <textarea value={this.state.value} onChange={this.handleChange}>
            Type Command [enter]
          </textarea>
        </form>
        <button className="btn btn-secondary" onClick={this.SubmitGameCommand}>
          Submit Command
        </button>
      </div>
    );
  }
}

export default GameConsole;
