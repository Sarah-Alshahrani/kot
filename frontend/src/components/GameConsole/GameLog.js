import React from "react";
import "./GameConsole";

class GameLog extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      username: props.currentUser,
      gameRoom: props.currentRoom,
      log: props.data
    };
  }

  componentDidMount() {
    this.scrollToBottom();
  }

  componentDidUpdate() {
    this.scrollToBottom();
  }

  scrollToBottom() {
    const chat = this.messagesEnd;

    if (chat) {
      const scrollHeight = chat.scrollHeight;
      const height = chat.clientHeight;
      const maxScrollTop = scrollHeight - height;
      chat.scrollTop = maxScrollTop > 0 ? maxScrollTop : 0;
    }
  }

  renderMessages(messages) {
    if (messages && messages.length > 0) {
      const currentUser = this.state.username;
      return messages.map((message, i) => (
        <li key={i} className={message.user === currentUser ? "me" : "him"}>
          <p>{message}</p>
        </li>
      ));
    }
  }

  render() {
    const game_messages = this.props.data;
    return (
      <div className="container">
        <ul
          ref={el => {
            this.messagesEnd = el;
          }}
        >
          {game_messages && this.renderMessages(game_messages)}
        </ul>
      </div>
    );
  }
}

export default GameLog;
