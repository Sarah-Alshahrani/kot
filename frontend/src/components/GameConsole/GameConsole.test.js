import React from "react";
import { shallow } from "enzyme";

import GameConsole from "./GameConsole";

describe("Verify GameConsole layout template", () => {
  it("GameConsole should render correctly", () => {
    const errorPage = shallow(<GameConsole />);
    expect(errorPage).toMatchSnapshot();
  });
});

describe("<GameConsole />", () => {
  const expected_user = "user_name";
  const expected_room = "room_name";
  const expected_cmd = "something_commandish";

  const testValues = {
    currentUser: expected_user,
    currentRoom: expected_room,
    cmd: expected_cmd,
    sendMessage: jest.fn(),
    hideMudUi: false
  };

  it("Submit works", () => {
    const component = shallow(<GameConsole {...testValues} />);

    component.setState({
      log: [],
      hideMudUi: false
    });

    const button = component.find("#send_button");
    expect(button.length).toBe(1);
    button.simulate("click");
    expect(testValues.sendMessage).toHaveBeenCalled();
    expect(testValues.sendMessage).toBeCalledWith({
      command: "gamelog_send_request",
      user: expected_user,
      room: expected_room,
      payload: expected_cmd
    });
  });
});
