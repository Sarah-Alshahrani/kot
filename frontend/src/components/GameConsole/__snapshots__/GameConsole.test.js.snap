// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify GameConsole layout template GameConsole should render correctly 1`] = `
<div
  className="game_console"
>
  <GameLog
    data={Array []}
  />
  <form
    className="submit_game_command"
    onChange={[Function]}
    onSubmit={[Function]}
  >
    <textarea
      onChange={[Function]}
    />
  </form>
  <button
    className="btn btn-secondary"
    onClick={[Function]}
  >
    Submit Command
  </button>
</div>
`;
